--- a/drivers/video/fbdev/sunxi/disp2/disp/fb_g2d_rot.c
+++ b/drivers/video/fbdev/sunxi/disp2/disp/fb_g2d_rot.c
@@ -24,13 +24,12 @@
 
 int fb_g2d_rot_free(struct fb_g2d_rot_t *inst)
 {
-	struct file g2d_file;
-
 	if (inst) {
+		if (inst->g2d_opened)
+			g2d_release(0, &inst->g2d_file);
 		disp_free((void *__force)inst->dst_vir_addr,
 		  (void *)inst->dst_phy_addr, inst->dst_mem_len);
 		kfree(inst);
-		g2d_release(0, &g2d_file);
 	}
 	return 0;
 }
@@ -69,9 +68,9 @@
 				inst->info.dst_image_h.height =
 				    inst->fb->var.xres;
 				inst->info.dst_image_h.clip_rect.w =
+				    inst->info.dst_image_h.width;
+				inst->info.dst_image_h.clip_rect.h =
 				    inst->info.dst_image_h.height;
-				inst->info.dst_image_h.clip_rect.h =
-				    inst->info.dst_image_h.width;
 				config->info.fb.crop.width =
 				    ((long long)inst->fb->var.yres << 32);
 				config->info.fb.crop.height =
@@ -200,6 +199,12 @@
 			inst->info.dst_image_h.clip_rect.x, inst->info.dst_image_h.clip_rect.x,
 			inst->info.dst_image_h.clip_rect.w, inst->info.dst_image_h.clip_rect.h);
 */
+	if (!inst->g2d_opened) {
+		if (g2d_open(0, &inst->g2d_file))
+			return -1;
+		inst->g2d_opened = true;
+	}
+
 	g2d_ioctl_mutex_lock();
 	ret = g2d_blit_h(&inst->info);
 	if (ret)
@@ -234,7 +239,6 @@
 {
 	int ret = -1;
 	struct fb_g2d_rot_t *fb_rot = NULL;
-	struct file g2d_file;
 	int degree;
 
 	if (!p_info || !config) {
@@ -254,10 +258,6 @@
 		pr_warn("kmalloc fail!!\n");
 		return NULL;
 	}
-	ret = g2d_open(0, &g2d_file);
-	if (ret)
-		goto ERROR;
-
 	fb_rot->info.src_image_h.width = p_info->var.xres;
 	fb_rot->info.src_image_h.height = p_info->var.yres;
 
@@ -318,7 +318,7 @@
 	fb_rot->dst_vir_addr =
 		disp_malloc(fb_rot->dst_mem_len, (u32 *)(&fb_rot->dst_phy_addr));
 	if (!fb_rot->dst_vir_addr || !fb_rot->dst_phy_addr)
-		goto G2D_RELEASE;
+		goto ERROR;
 	config->info.fb.addr[0] =
 	    (unsigned long long)fb_rot->dst_phy_addr + fb_rot->dst_mem_len / 2;
 
