--- a/drivers/video/fbdev/sunxi/disp2/disp/fb_g2d_rot.c	2026-05-18 20:20:21.152813407 +0000
+++ b/drivers/video/fbdev/sunxi/disp2/disp/fb_g2d_rot.c	2026-05-18 20:20:08.662825302 +0000
@@ -159,6 +159,7 @@
 
 int fb_g2d_rot_apply(struct fb_g2d_rot_t *inst, struct disp_layer_config *config)
 {
+	return 0;  /* DIAG 009: early return to isolate hang */
 	int ret = -1;
 
 	if (!inst || !config || !inst->fb) {
